### PR TITLE
refactoring for efficiency

### PR DESCRIPTION
**Pull request by alsgregory**
Alterations:

- Changed a 1d demo
- Altered flux, slope_modification, slope_limiter and timestepper to refactor time-stepping of model
- This is more efficient than currently on master, given that it just creates a set-up solver and then updates this solver for future time-steps

### CODE REVIEW

File: flooddrake/timestepper.py [hunk 1]
```diff
@@ -211,53 +215,49 @@ def stepper(self, t_start, t_end, w):
...
         bout_file.write(bout)
 
         self.t = 0
 
         for i in range(Nt):
 
-            w_ = self.__solver_variable(w)
+            solve(self.L == 0, self.w_, nest=False,
+                  solver_parameters={'ksp_type': 'preonly',
+                                     'pc_type': 'lu'})
```
Comment (wence-) [inline flooddrake/timestepper.py]: This is a nonlinear solve.  Is that actually what you wanted?  I also recommend that you pull the solver out of the time loop:

```
problem = NonlinearVariationalProblem(self.L, self._w)
solver = NonlinearVariationalSolver(problem, ...)

for i in range(Nt):
    solver.solve()
    self.w.assign(self.w_)
    self.__update_slope_modification()
    solver.solver()
    ...
```

File: flooddrake/timestepper.py [hunk 2]
```diff
@@ -211,53 +215,49 @@ def stepper(self, t_start, t_end, w):
...
-                # index functions and spaces
-                h, mu = split(w)
-
-            if self.mesh.geometric_dimension() == 2:
-
-                # index functions and spaces
-                h, mu, mv = split(w)
-
-            hout_file.write(hout.project(h + self.b_))
+            hout_file.write(hout.project(self.h + self.b_))
```
Comment (wence-) [inline flooddrake/timestepper.py]: You could make a `Projector` outside the time loop and use that here.

File: flooddrake/timestepper.py (hunk 1 above)
Comment (alsgregory) [inline flooddrake/timestepper.py]: Thanks, fixed this. This is now A LOT faster!

File: flooddrake/timestepper.py [hunk 3]
```diff
@@ -100,74 +117,69 @@ def __solver_variable(self, w):
...
+                      dot(self.v.dx(0), self.F) * dx +
+                      (dot(self.v('-'), self.NegFlux) + dot(self.v('+'), self.PosFlux)) * dS +
+                      dot(self.v, self.BoundaryFlux) * ds +
+                      (dot(self.source, self.v)) * dx -
+                      (dot(self.v('-'), self.delta_minus) +
+                      dot(self.v('+'), self.delta_plus)) * dS)
 
-        return w_
+        self.problem = NonlinearVariationalProblem(self.L, self.w_)
+        self.solver = NonlinearVariationalSolver(self.problem)
```
Comment (wence-) [inline flooddrake/timestepper.py]: You'll want to set appropriate solver parameters here:

```python
NonlinearVariationalSolver(self.problem, solver_parameters={...})
```

Note that `nest=False` lives with the `Problem`, not the `Solver`.

**Comment by alsgregory** [comment]
Can I merge now tests have passed?

**Comment by colinjcotter** [comment]
Yes, please merge

**Pull request merged**